### ISSUE TITLE
Highlight groups not properly cleared ("Too many highlight and syntax groups")

### ISSUE THREAD

**Issue by Edenhofer** [neovim/neovim#16494]
### Neovim version (nvim -v)

NA commit 73b35ef10f955 (git describe: v0.5.0-1213-g73b35ef10)

### Vim (not Nvim) behaves the same?

no, vim 8.2.3582

### Operating system/version

Arch Linux

### Terminal name/version

kitty 0.23.1

### $TERM environment variable

tmux-256color

### Installation

Arch User Repository (neovim-git)

### How to reproduce the issue

The error is pretty generic, see the below explanation. It can be easily reproduced using the following `.vimrc` if one provokes many changes that update highlighting groups.

```vimrc
" This line should not be removed as it ensures that various options are
" properly set to work with the Vim-related packages.
runtime! archlinux.vim
" Move around code blocks in e.g. Julia
runtime macros/matchit.vim

call plug#begin('~/.vim/plugged')
Plug 'mhinz/vim-signify'  " VCS sign indicators
call plug#end()
```

### Expected behavior

No errors regardless of how long I keep my vim session open.

### Actual behavior

After a while errors similar to this one `E849: Too many highlight and syntax groups` pop up and all new highlighting in vim stops working. This can be further confirmed by looking at the highlighting groups (`:hi`) which contain thousands of the following lines:
```
           xxx cleared
           xxx cleared
           xxx cleared
           xxx cleared
           xxx cleared
```

The problem seems to be that syntax groups (here induced by vim-signify but any plugin that frequently updates highlighting groups breaks nvim) are not properly cleared and nvim eventually runs into the limit specified at [nvim/syntax.c](https://github.com/neovim/neovim/blob/73b35ef/src/nvim/syntax.c#L101). This does not happend with nvim-v0.4.3 thus I assume it is a bug in nvim rather than the various plugins that update highlighting groups.

**Comment by zeertzjq** [comment]
What highlight group names do you see in the thousands of lines?

**Comment by maxxnino** [comment]
Lunarvim have it too, and they solved the issue
https://github.com/LunarVim/LunarVim/issues/1880

**Comment by Edenhofer** [comment]
Sorry if this has not been clear, I see no highlight group names. The above example is a snippet from the full output of `:hi`. To make the above output more clear, here is another excerpt from the output of `:hi`:
```
SignifySignDeleteFirstLine xxx links to SignifySignDelete
SignifySignChange xxx links to DiffChange
SignifySignChangeDelete xxx links to SignifySignChange
               xxx cleared
               xxx cleared
               xxx cleared
               xxx cleared
```

**Comment by zeertzjq** [comment]
Seems to be related to signs.

**Comment by Edenhofer** [comment]
Maybe, but I am nor sure about this. I see the very same error also with other plugin that induce many highlighting groups.

**Comment by Edenhofer** [comment]
The bugs reference by @maxxnino seem very much related! Thanks for referencing them!

**Comment by zeertzjq** [comment]
Minimum repro vimrc:
```vim
function! Repro()
  sign define ReproSign text=R texthl=Operator linehl=Operator
  sign define ReproSign text=R texthl=Operator linehl=
endfunction

nnoremap <Space> <Cmd>call Repro()<CR>
```

Run `nvim -u` with this file, hold down space bar, and there will be a lot of these:
```
           xxx cleared
           xxx cleared
           xxx cleared
           xxx cleared
           xxx cleared
```

**Comment by zeertzjq** [comment]
This bug is not present in v0.5.1.

**Comment by zeertzjq** [comment]
Bisected to <https://github.com/neovim/neovim/commit/a9f563ab62f6035729f79c3c701a5b219f1b982f>.

**Comment by erw7** [comment]
This is indeed a problem introduced in a9f563a, but I think this problem is a manifestation of an upstream problem where highlight groups with empty strings in their names are created. I created a vim/vim#9280 about it. If this is not fixed upstream, the following condition needs to be modified to `if (len > 200) {`.

https://github.com/neovim/neovim/blob/1e6eeca9d1360554ee18525603e83c3a1999a622/src/nvim/syntax.c#L7628

**Comment by zeertzjq** [comment]
Bram chose to make empty highlight group with `:sign` fail. <https://github.com/vim/vim/commit/5e18ccc60bdddc4aa39ab039f1a7c918f29e67ce>

**Comment by jamessan** [comment]
> Bram chose to make empty highlight group with `:sign` fail in [vim/vim@5e18ccc](https://github.com/vim/vim/commit/5e18ccc60bdddc4aa39ab039f1a7c918f29e67ce).

That's been reverted now.

> If this is not fixed upstream, the following condition needs to be modified to `if (len > 200) {`.

I think that's the right fix.

**Comment by zeertzjq** [comment]
#16541 seems to fix this.

**Comment by erw7** [comment]
>>Bram chose to make empty highlight group with :sign fail in vim/vim@5e18ccc.
>
>That's been reverted now. 

vim/vim@e5710a0 only partially reverts vim/vim@5e18ccc. Therefore, this issue is resolved by #16541.

**Comment by erw7** [comment]
The same problem occurs with name highlight groups that are longer than the buffer of `syn_name2id`(`syn_name2id_len`). For this reason I created vim/vim#9289.
